File: paleo_workbench/catalog/lineage_graph.py
"""Full-chain lineage walks over the canonical document (collaborator module).

Two read-only services used by :class:`DataCatalogService`:

- :func:`build_lineage_chain` — the recursive version→run→version chain from
  any version up to its RAW roots (or down to its descendants), as a tree of
  :class:`LineageChainNode` interleaved with producing-run info. This powers
  the Data Manager's 血缘 visualization: one call answers "which seismic /
  horizon / well inputs, through which runs, produced this output".
- :func:`compute_summaries` — per-version lineage status (hops to the nearest
  RAW ancestor, broken-parent flag) in one pass, memoized inside DFS so the
  diamond-heavy graphs stay cheap. Powers the table's 血缘 column.

Both walk the service's maintained id maps (rebuilt once per revision, not
per query) and are cycle-safe via visited sets.
"""
from __future__ import annotations

from collections import deque
from dataclasses import dataclass, field
from typing import Any

from paleo_workbench.catalog.models import DataStage

DEFAULT_MAX_NODES = 5000


@dataclass
class LineageChainNode:
    """One version in the lineage tree plus its producing-run facts."""

    version_id: str
    asset_id: str
    asset_name: str
    stage: DataStage
    version_number: int
    depth: int  # hops from the start version
    managed: bool = True
    trashed: bool = False
    path: str = ""
    sha256: str | None = None
    created_at: str = ""
    tags: list[str] = field(default_factory=list)
    run_id: str | None = None
    run_operation: str | None = None
    run_status: str | None = None
    run_generator: str | None = None
    children: list["LineageChainNode"] = field(default_factory=list)


@dataclass
class LineageChain:
    """Result of :func:`build_lineage_chain`.

    ``root`` is the start version; ancestors mode expands ``children`` towards
    RAW inputs, descendants mode towards downstream products. ``truncated`` is
    True when the walk hit *max_depth* / *max_nodes* (the display should say
    so instead of implying the chain ends).
    """

    start_version_id: str
    direction: str  # "ancestors" | "descendants"
    root: LineageChainNode
    node_count: int = 1
    truncated: bool = False


def _version_tags(tag_by_id: dict[str, Any], version_tags: Any, version_id: str) -> list[str]:
    try:
        tag_ids = version_tags.get(version_id, [])
        return [
            tag_by_id[tid].display_name or tag_by_id[tid].name
            for tid in tag_ids
            if tid in tag_by_id
        ]
    except Exception:
        return []


def _make_node(
    maps: Any,
    tag_by_id: dict[str, Any],
    version_tags: Any,
    version: Any,
    depth: int,
) -> LineageChainNode:
    try:
        asset_name = maps.asset_by_id[version.asset_id].name
    except Exception:
        asset_name = version.asset_id
    run = maps.run_by_id.get(version.run_id) if version.run_id is not None else None
    return LineageChainNode(
        version_id=version.id,
        asset_id=version.asset_id,
        asset_name=asset_name,
        stage=version.stage,
        version_number=version.version_number,
        depth=depth,
        managed=version.managed,
        trashed=version.trashed,
        path=version.path,
        sha256=version.sha256,
        created_at=version.created_at or "",
        tags=_version_tags(tag_by_id, version_tags, version.id),
        run_id=version.run_id,
        run_operation=run.operation if run is not None else None,
        run_status=run.status if run is not None else None,
        run_generator=run.generator if run is not None else None,
    )


def build_lineage_chain(
    service: Any,
    version_id: str,
    *,
    direction: str = "ancestors",
    max_depth: int | None = None,
    max_nodes: int = DEFAULT_MAX_NODES,
) -> LineageChain:
    """Walk the full lineage tree from *version_id* (cycle-safe).

    ``direction="ancestors"`` follows ``parent_version_ids`` towards RAW
    inputs; ``"descendants"`` follows the maintained child index towards
    downstream products. Because parents mirror run inputs (the adapter's
    ``_register_produced`` convention), the ancestor tree IS the
    version→run→input chain; each node carries its producing run for display.
    """
    if direction not in ("ancestors", "descendants"):
        raise ValueError(f"direction must be 'ancestors' or 'descendants', got {direction!r}")
    with service._lock:
        start = service._version_or_raise(version_id)
        maps = service._ensure_maps()
        by_id = maps.version_by_id
        children_index = maps.children_by_parent
        # One precomputed index set per traversal (#1059): rebuilding the tag
        # map per node cost O(V×T) and get_asset/get_run re-locked per lookup.
        tag_by_id = {t.id: t for t in service.document.tags}
        version_tags = service.document.version_tags

        root = _make_node(maps, tag_by_id, version_tags, start, 0)
        seen = {start.id}
        truncated = False
        queue = deque([root])
        while queue:
            node = queue.popleft()
            version = by_id.get(node.version_id)
            if version is None:
                continue
            if direction == "ancestors":
                next_ids: list[str] = list(version.parent_version_ids)
            else:
                next_ids = [v.id for v in children_index.get(node.version_id, ())]
            if max_depth is not None and node.depth >= max_depth:
                # Only a real truncation when there was more to expand.
                truncated = truncated or bool(next_ids)
                continue
            for next_id in next_ids:
                child_version = by_id.get(next_id)
                if child_version is None or next_id in seen:
                    continue  # dangling parents are audit findings, not tree nodes
                if len(seen) >= max_nodes:
                    truncated = True
                    break
                seen.add(next_id)
                child = _make_node(maps, tag_by_id, version_tags, child_version, node.depth + 1)
                node.children.append(child)
                queue.append(child)
        return LineageChain(
            start_version_id=start.id,
            direction=direction,
            root=root,
            node_count=len(seen),
            truncated=truncated,
        )


def compute_summaries(service: Any) -> dict[str, dict[str, Any]]:
    """Per-version lineage status in one pass:

    ``{"to_raw": int | None, "broken": bool, "has_parents": bool}`` where
    ``to_raw`` is the minimum hop count to a RAW ancestor (0 when the version
    itself is RAW; None when no RAW ancestor is reachable — e.g. an isolated
    derived registration) and ``broken`` flags dangling parent references.

    Iterative DFS with memoization; a cycle terminates at the visited set and
    reports the cycle member as unreachable-from-there (audit reports the
    cycle itself separately).
    """
    with service._lock:
        by_id = service._ensure_maps().version_by_id
        memo: dict[str, int | None] = {}
        broken: set[str] = set()
        has_parents: dict[str, bool] = {}

        def visit(start_id: str) -> int | None:
            # Iterative post-order DFS: (version_id, phase) with phase 0 =
            # enter, 1 = merge children.
            stack: list[tuple[str, int]] = [(start_id, 0)]
            on_path: set[str] = set()
            local_results: dict[str, int | None] = {}
            while stack:
                vid, phase = stack.pop()
                if vid in memo:
                    local_results[vid] = memo[vid]
                    continue
                version = by_id.get(vid)
                if version is None:
                    local_results[vid] = None
                    memo[vid] = None
                    continue
                if phase == 0:
                    if vid in on_path:
                        # Cycle: treat as no-path-through-here (audit reports it).
                        local_results[vid] = None
                        continue
                    if not version.parent_version_ids:
                        memo[vid] = 0 if version.stage == DataStage.RAW else None
                        local_results[vid] = memo[vid]
                        continue
                    on_path.add(vid)
                    stack.append((vid, 1))
                    for pid in version.parent_version_ids:
                        if pid not in by_id:
                            broken.add(vid)
                        stack.append((pid, 0))
                    continue
                # phase 1: children evaluated into local_results/memo
                on_path.discard(vid)
                best: int | None = None
                if version.stage == DataStage.RAW:
                    best = 0
                for pid in version.parent_version_ids:
                    child_depth = local_results.get(pid, memo.get(pid))
                    if child_depth is None:
                        continue
                    candidate = child_depth if version.stage == DataStage.RAW else child_depth + 1
                    best = candidate if best is None else min(best, candidate)
                memo[vid] = best
                local_results[vid] = best
            return local_results.get(start_id)

        summaries: dict[str, dict[str, Any]] = {}
        for version in service.document.versions:
            has_parents[version.id] = bool(version.parent_version_ids)
            visit(version.id)
        for version in service.document.versions:
            summaries[version.id] = {
                "to_raw": memo.get(version.id),
                "broken": version.id in broken,
                "has_parents": has_parents[version.id],
            }
        return summaries

File: libs/catalog/src/lineage_graph.cpp
#include "pwb/catalog/lineage_graph.hpp"

#include <deque>
#include <set>

namespace pwb::catalog {

namespace {

std::map<std::string, std::vector<std::string>> version_tag_ids(
    const CatalogDocument& document) {
    std::map<std::string, std::vector<std::string>> out;
    for (const auto& [owner, tag_id] : document.version_tags) {
        out[owner].push_back(tag_id);
    }
    return out;
}

LineageChainNode make_node(const CatalogDocument& document,
                           const DocumentIndex& index, const DataVersion& version,
                           int depth,
                           const std::map<std::string, std::vector<std::string>>& tags_by_owner) {
    LineageChainNode node;
    node.version_id = version.id.str();
    node.asset_id = version.asset_id.str();
    const DataAsset* asset = index.asset(node.asset_id);
    node.asset_name = asset != nullptr ? asset->name : node.asset_id;
    node.stage = version.stage;
    node.version_number = version.version_number;
    node.depth = depth;
    node.managed = version.managed;
    node.trashed = version.trashed;
    node.path = version.path;
    node.sha256 = version.sha256;
    node.created_at = version.created_at;
    auto tags = tags_by_owner.find(node.version_id);
    if (tags != tags_by_owner.end()) {
        for (const auto& tag_id : tags->second) {
            for (const auto& tag : document.tags) {
                if (tag.id == tag_id) {
                    node.tags.push_back(tag.display_name.value_or(tag.name));
                    break;
                }
            }
        }
    }
    node.run_id = version.run_id.has_value()
                      ? std::optional<std::string>(version.run_id->str())
                      : std::nullopt;
    if (version.run_id.has_value()) {
        const DataRun* run = index.run(version.run_id->str());
        if (run != nullptr) {
            node.run_operation = run->operation;
            node.run_status = run->status;
            node.run_generator = run->generator;
        }
    }
    return node;
}

}  // namespace

domain::Result<LineageChain> build_lineage_chain(
    const CatalogDocument& document, const DocumentIndex& index,
    const std::string& version_id, const std::string& direction,
    std::optional<int> max_depth, int max_nodes) {
    if (direction != "ancestors" && direction != "descendants") {
        // ValueError text parity: f"direction must be 'ancestors' or
        // 'descendants', got {direction!r}"
        std::string repr = "'" + direction + "'";
        return domain::DataError(
            domain::ErrorCode::InvalidArgument,
            "direction must be 'ancestors' or 'descendants', got " + repr);
    }
    const DataVersion* start = index.version(version_id);
    if (start == nullptr) {
        return domain::DataError(domain::ErrorCode::NotFound,
                                 "Unknown version: " + version_id);
    }
    const auto tags_by_owner = version_tag_ids(document);

    LineageChain chain;
    chain.start_version_id = start->id.str();
    chain.direction = direction;
    chain.root = make_node(document, index, *start, 0, tags_by_owner);

    std::set<std::string> seen{start->id.str()};
    std::deque<LineageChainNode*> queue;
    queue.push_back(&chain.root);
    bool truncated = false;
    while (!queue.empty()) {
        LineageChainNode* node = queue.front();
        queue.pop_front();
        const DataVersion* version = index.version(node->version_id);
        if (version == nullptr) continue;
        std::vector<std::string> next_ids;
        if (direction == "ancestors") {
            for (const auto& parent : version->parent_version_ids) {
                next_ids.push_back(parent.str());
            }
        } else {
            if (const auto* children = index.children_of(node->version_id)) {
                for (const DataVersion* child : *children) {
                    next_ids.push_back(child->id.str());
                }
            }
        }
        if (max_depth.has_value() && node->depth >= *max_depth) {
            truncated = truncated || !next_ids.empty();
            continue;
        }
        for (const auto& next_id : next_ids) {
            const DataVersion* child_version = index.version(next_id);
            if (child_version == nullptr || seen.count(next_id)) continue;
            if (static_cast<int>(seen.size()) >= max_nodes) {
                truncated = true;
                break;
            }
            seen.insert(next_id);
            node->children.push_back(
                make_node(document, index, *child_version, node->depth + 1, tags_by_owner));
            queue.push_back(&node->children.back());
        }
    }
    chain.node_count = static_cast<int>(seen.size());
    chain.truncated = truncated;
    return chain;
}

std::map<std::string, LineageSummary> compute_lineage_summaries(
    const CatalogDocument& document, const DocumentIndex& index) {
    std::map<std::string, std::optional<int>> memo;
    std::set<std::string> broken;
    std::map<std::string, LineageSummary> summaries;

    for (const auto& version : document.versions) {
        const std::string start_id = version.id.str();
        if (memo.count(start_id)) continue;
        // Iterative post-order DFS: (id, phase) with phase 0 = enter,
        // 1 = merge children — the Python structure verbatim.
        std::vector<std::pair<std::string, int>> stack{{start_id, 0}};
        std::set<std::string> on_path;
        std::map<std::string, std::optional<int>> local_results;
        while (!stack.empty()) {
            auto [vid, phase] = stack.back();
            stack.pop_back();
            if (memo.count(vid)) {
                local_results[vid] = memo[vid];
                continue;
            }
            const DataVersion* version_ptr = index.version(vid);
            if (version_ptr == nullptr) {
                local_results[vid] = std::nullopt;
                memo[vid] = std::nullopt;
                continue;
            }
            if (phase == 0) {
                if (on_path.count(vid)) {
                    // Cycle: treat as no-path-through-here.
                    local_results[vid] = std::nullopt;
                    continue;
                }
                if (version_ptr->parent_version_ids.empty()) {
                    memo[vid] = version_ptr->stage == domain::DataStage::Raw
                                    ? std::optional<int>(0)
                                    : std::nullopt;
                    local_results[vid] = memo[vid];
                    continue;
                }
                on_path.insert(vid);
                stack.push_back({vid, 1});
                for (const auto& pid : version_ptr->parent_version_ids) {
                    if (index.version(pid.str()) == nullptr) {
                        broken.insert(vid);
                    }
                    stack.push_back({pid.str(), 0});
                }
                continue;
            }
            on_path.erase(vid);
            std::optional<int> best;
            if (version_ptr->stage == domain::DataStage::Raw) best = 0;
            for (const auto& pid : version_ptr->parent_version_ids) {
                std::optional<int> child_depth;
                auto local = local_results.find(pid.str());
                if (local != local_results.end()) {
                    child_depth = local->second;
                } else {
                    auto global = memo.find(pid.str());
                    if (global != memo.end()) child_depth = global->second;
                }
                if (!child_depth.has_value()) continue;
                int candidate = version_ptr->stage == domain::DataStage::Raw
                                    ? *child_depth
                                    : *child_depth + 1;
                if (!best.has_value() || candidate < *best) best = candidate;
            }
            memo[vid] = best;
            local_results[vid] = best;
        }
    }
    for (const auto& version : document.versions) {
        LineageSummary summary;
        auto it = memo.find(version.id.str());
        summary.to_raw = it != memo.end() ? it->second : std::nullopt;
        summary.broken = broken.count(version.id.str()) != 0;
        summary.has_parents = !version.parent_version_ids.empty();
        summaries[version.id.str()] = summary;
    }
    return summaries;
}

}  // namespace pwb::catalog
